www/source/blog/2018-08-18-lift-and-shift-hab.html.md

```diff
@@ -0,0 +1,396 @@
+
+````bash
+$ git clone https://github.com/skpaterson/table-setting
+$ cd table-setting
+$ python -m venv tsenv
+$ . tsenv/bin/activate
+$ pip install -r requirements.txt
+$ python run_app.py 
+````
+
+Navigating to [http://localhost:5000](http://localhost:5000) you should see the Table Setting Swagger UI:
+
+![Running Local InMemory SQLite](media/2018-08-18-lift-and-shift-hab/local_in_memory.png)
+
+Feel free to try creating some forks!  By default this runs with a clean SQLite in-memory database that is wiped on restart and fork CRUD operations are available. 
+
+A user running even this simple application locally is perhaps unnecessarily exposed to several things:
+* Source control with `git`
+* Python language installation
+* Virtual environment setup
+* Python packaging with `pip` 
+* Internal application structure e.g. where to find the thing to run
+
+A similar list could likely be formulated for applications written in other languages.  In any case, this complicates life for performing any changes at a later stage e.g.
+* Switching to a MySQL database backend
+* Running on a new machine, or
+* Moving to a (new) Cloud provider or service
+
+
+## Packaging Table Setting with Habitat
+
+The following directory structure within the Table Setting repository relates to Habitat packaging:
+
+```bash
+$ tree habitat/
+habitat/
+├── README.md
+├── config
+├── default.toml
+├── hooks
+│   ├── init
+│   └── run
+└── plan.sh
+```
+This incorporates a set of steps to install required dependencies, similar to those taken when running locally and also includes the MySQL library setup.  Table Setting explicitly depends on a known good Python version. Environment variables are used to convey runtime configuration to Table Setting, some are conditionally set when e.g. binding to a Habitat managed MySQL database.  
+
+Let's concentrate for now on creating deployment artifacts with Habitat.  Below we use the Studio to export different package formats:
+```bash
+$ cd table-setting
+$ hab studio enter
+[STUDIO] build
+[STUDIO] hab pkg export tar $(ls -1t results/*.hart | head -1)
+[STUDIO] hab pkg export docker $(ls -1t results/*.hart | head -1)
+```
+
+We also use the Habitat maintained `core/mysql` image for some deployments, let's install and export that locally also:
+
+```bash
+[STUDIO] hab pkg install core/mysql
+» Installing core/mysql
+☁ Determining latest version of core/mysql in the 'stable' channel
+☛ Verifying core/mysql/5.7.21/20180609181124
+... output curtailed
+★ Install of  core/mysql/5.7.21/20180609181124 complete with 25 new packages installed.
+```
+
+Note the package version from the installation command, we will now use this to export a Docker image locally:
```

Comment:
"Note the package version from the installation command (the latest version may be different when you run this); we will now use this to export a Docker image locally:"

Note, however, that you can just `hab pkg export docker core/mysql` without previously installing, and you'll get the latest stable package automatically.

Reply:
👍